File: python/23.py
# problem 23
# find the sum of the all positive integers which cannot be written as the sum of two abundant numbers

def abundant(num):
  """this function will find the proper divisors of a number"""
  factors = []
  result = 0
  for factor in reversed(range(1, int(num/2)+1)):
    if num % factor == 0:
      factors.append(factor)
  for each in factors: result+=each
  return result > num
    
print("find all the abundant numbers...")

isAbundant = []
result = 0
# find all the abundant numbers under 28123
for each in range(1, 28124):
  result += each
  if abundant(each):
    isAbundant.append(each)

print("found all the abundant numbers...")

# TODO make this not broken
counter = 1
for number in isAbundant:
  # lets see which ones it could be
  # if it's abundant to start with, ignore it.
  for possibility in isAbundant[counter:]:
    index = number - possibility
    if index < 28124:
      result-=isAbundant[counter]
  counter+=1
print(result)

File: ruby/23.rb
require "set"

class Integer
  def abundant?
    result = 0
    Integer(self/2+1).downto(1) do |factor|
      if self % factor == 0
        # factors.append(factor)
        result += factor
      end
    end
    return result > self
  end
end

puts 'Finding all the abundant numbers...'

isAbundant = Set.new
result = 0

(3..28123).each do |number|
  if number.abundant?
    # print "#{number}\n"
    isAbundant.add(number)
  end
end

print "found all the abundant numbers... (#{isAbundant.size})\n"

possibilities = *(0..28123)

isAbundant.each do |factorOne|
  isAbundant.each do |factorTwo|
    result = factorOne+factorTwo
      # puts "We've decided that #{result} is not valid for our purposes #{possibilities[result]}"
      possibilities[result]=0
  end
end

# print sum of the remaining possibilities
possibilities.compact!
puts possibilities.inject(:+)
